Correct handling of min/maxdepth during file.find (#34217)

anderbubble opened the pull request:
### What does this PR do?

Fixes handling of min/maxdepth during `file.find`
### What issues does this PR fix or reference?
#34217
### Previous Behavior

Specifying min/maxdepth during `file.find` included and excluded the incorrect paths.
### New Behavior

Specifying min/maxdepth during `file.find` includes and excludes the same paths that unix find would.
### Tests written?

Yes.

anderbubble committed:
Correct handling of min/maxdepth during file.find (#34217)

cachedout commented on salt/utils/find.py:
```diff
@@ -655,10 +659,22 @@ def find(self, path):
+def path_depth(path):
+    depth = 0
+    head = path
+    while True:
```
This looks like something that can infinitely recurse. If we don't hit the break on the first iteration, what causes this not to loop endlessly?

anderbubble replied:
There is no recursion; only iteration.

The break condition variable, `tail`, is assigned on every iteration, on line 670.

`head, tail = os.path.split(head)`

`head` is iteratively deconstructed, with the next `tail` removed each time. So `head` iterates to `/` or `.`, while `tail` iterates to `.` or ``.

I did find a bug, though, in the case where path ends in `/`. I'll fix that.

cachedout replied:
Yes, iteration. Sorry. I'm running on very little sleep today. Thanks.

anderbubble committed:
Fix salt.util.path_depth where path ends in '/'

salt.util.path_depth uses os.path.split, which handles input
paths differently when they end in '/' (or os.path.sep in
general). This handles that case.

Note that this behavior doesn't appear to impact file.find,
as the argument to path_depth passes through os.path.relpath
first. Still, it's better to have the path_depth behavior
correct for all potential inputs.

anderbubble committed:
Unify components ignored by salt.util.path_depth

cachedout commented:
Hi @anderbubble 

It does look like this test is now failing: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/2551/testReport/junit/unit.utils.find_test/TestFinder/test_find/

Commit:
Addressed unit testing for salt.utils.find

In addition to addressing test failures introduced by this
bugfix branch, I've also added additional tests for the
proper behavior of min/maxdepth.

anderbubble commented:
I addressed the failing unit tests, and I added some new tests for the correct behavior of min/maxdepth.

In addressing the failing tests, I ended up refactoring two algorithms from `Finder.find` into distinct methods. This does mean that there are more function calls in the critical path; but I think it's worth it for the sake of code clarity.

I also now calculate `fullpath` in all cases, since it's always calculated for matches anyway. I could see an argument that this incurs some performance penalty for pervasively non-matching finds; but I can't imagine that `os.path.join` is *that* inefficient, especially compared to the development overhead of maintaining it as a condition, and the functional overhead of repeatedly evaluating its presence.

I *have* maintained the memoization of `fstat`, however, as fs io is potentially vastly more expensive than string manipulation.

anderbubble commented:
@cachedout I see that new commits to the develop branch now conflict with this change, due to a fix in how this function handles symlinks. Is it your preference that I rebase this pull request on top of the new develop, or will you/upstream handle the merge conflict?

Besides that, anything else you need from me to get this merged?

rallytime commented:
@anderbubble If you could rebase this PR on top of new develop, that would be fantastic.

anderbubble commented:
@rallytime No problem. I've done the rebase, and (force) pushed to the PR branch. Let me know if you need anything else for the merge.

cachedout commented:
@anderbubble Just one minor lint error and then we can likely get this in: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/2874/violations/file/salt/utils/find.py/

Thanks!

Commit:
Update for PEP8 E302 conformity

anderbubble commented:
@cachedout pylint error should be fixed now.

cachedout commented:
@anderbubble Thanks so much!
